### ISSUE TITLE
python: process is uninterruptible when in flux_reactor_run()

### ISSUE THREAD

**Issue by grondo** [flux-framework/flux-core#2707]
By default python scripts that call into `h.reactor_run()` are uninterruptible.

I haven't investigated further yet, but a simple reproducer is to just run the reactor with a timer registered:
```python
import flux

h = flux.Flux()
r = h.get_reactor()
h.timer_watcher_create (5., lambda h, w, x, y: h.reactor_stop(r)).start()
h.reactor_run(r, 0)
```

Running the above, SIGINT is blocked for while in `reactor_run`, and I have to ctrl-z or SIGQUIT:

```console
$ flux python test.py
^C^C^C^C^\Quit (core dumped)
```

In contrast, the following C test program is interruptible:
```c
#include <stdio.h>
#include <flux/core.h>

int main ()
{
	flux_t *h;
	flux_reactor_t *r;
	flux_watcher_t *w;
	if (!(h = flux_open (NULL, 0))) {
		fprintf (stderr, "flux_open failed\n");
		exit (1);
	}
	r = flux_get_reactor (h);
	if (!(w = flux_timer_watcher_create (r, 5., 0., NULL, NULL))) {
		fprintf (stderr, "Failed to create timer\n");
		exit (1);
	}
	flux_watcher_start (w);
	if (flux_reactor_run (r, 0) < 0) {
		fprintf (stderr, "flux_reactor_run() returned < 0 \n");
		exit (1);
	}
	exit (0);
}
```

**Comment by SteVwonder** [comment]
Hmm. I guessing the issue is that Python is receiving the SIGINT signal and the queueing up the handler to run after the C function returns.  That seems to be supported by [this Python documentation.](https://docs.python.org/3/library/signal.html#execution-of-python-signal-handlers)

I am not finding any satisfying answers in the `cffi` documentation as to how to handle this. Going to their issue tracker, I found [this](https://bitbucket.org/cffi/cffi/issues/306/provide-a-wrapper-for-_pyos_sigintevent-or), which lead to [this](https://github.com/python-trio/trio/issues/42), and eventually [this PR](https://github.com/python-trio/trio/pull/108/files) and [this Python documentation on the signal.set_wakeup_fd](signal.set_wakeup_fd) function.   I don't know if this is the best way, but I'm wondering if it would work to:

- use `signal.set_wakeup_fd` to create a file descriptor in Python that will be written to when `Ctrl-C` is pressed
- pass that fd down into the flux reactor with `flux_fd_watcher_create` and set the callback to be a python cffi callback
- call `flux_reactor_stop` or bail out in some other way from that callback

I'm a little worried that this won't be enough, and that we will have to create a timer that periodically kicks control back to Python to handle signals 🤢 .

One alternative, long-term solution would be to integrate the Py3 async event loop with Flux's reactor loop and use the python loop directly instead.

Any other thoughts or hacks @trws?

**Comment by trws** [comment]
I think the simplest thing would be to do something like what you suggest but probably just a c handler that python sets up to cause the reactor to exit. However we do it, I think the goal would be to get the reactor to at least temporarily exit on a signal like that, maybe even as a default behavior of the reactor? What normally happens in C if the user doesn’t register a handle, kill?

My plan for the event loop support has been to use the flux reactor as a replacement for epoll effectively, so that would likely hit the same issue if we don’t get a good handle on adding waits on signals either directly or through FDs.

Sent from my iPhone

On Feb 3, 2020, at 11:06 PM, Stephen Herbein <notifications@github.com> wrote:

﻿

Hmm. I guessing the issue is that Python is receiving the SIGINT signal and the queueing up the handler to run after the C function returns. That seems to be supported by this Python documentation.<https://docs.python.org/3/library/signal.html#execution-of-python-signal-handlers>

I am not finding any satisfying answers in the cffi documentation as to how to handle this. Going to their issue tracker, I found this<https://bitbucket.org/cffi/cffi/issues/306/provide-a-wrapper-for-_pyos_sigintevent-or>, which lead to this<https://github.com/python-trio/trio/issues/42>, and eventually this PR<https://github.com/python-trio/trio/pull/108/files> and this Python documentation on the signal.set_wakeup_fd<signal.set_wakeup_fd> function. I don't know if this is the best way, but I'm wondering if it would work to:

  *   use signal.set_wakeup_fd to create a file descriptor in Python that will be written to when Ctrl-C is pressed
  *   pass that fd down into the flux reactor with flux_fd_watcher_create and set the callback to be a python cffi callback
  *   call flux_reactor_stop or bail out in some other way from that callback

I'm a little worried that this won't be enough, and that we will have to create a timer that periodically kicks control back to Python to handle signals 🤢 .

One alternative, long-term solution would be to integrate the Py3 async event loop with Flux's reactor loop and use the python loop directly instead.

Any other thoughts or hacks @trws<https://github.com/trws>?

**Comment by grondo** [comment]
I was going to take a quick stab at a fix for this. It would be nice to have this fixed in our next release.

It would also be nice to change the current use of `reactor_run` from the currently required `handle.reactor_run(handle.get_reactor(), 0)` to simply `handle.reactor_run()`

What I'll try is to add a new `reactor_run` method to the `Flux` class. Unfortunately since `handle.reactor_run` is already in use, this might have to be `handle.reactor()`.

Once that is working, I'll see if I can get the default signal handling to work as mentioned above.

**Comment by grondo** [comment]
Proof of concept (after adding missing `SignalWatcher` class.):
```python
    def _interrupt(self):
        self.interrupted = True
        self.reactor_stop(self.get_reactor())

    def reactor(self, flags=0):
        import signal
        from flux.core.watchers import SignalWatcher
        sigint_watcher = SignalWatcher(
                self,
                signal.SIGINT,
                lambda h, w, r, y, x: h._interrupt()
        )   
        sigint_watcher.start()
        self.interrupted = False
        rc = self.flux_reactor_run(self.get_reactor(), flags)
        if self.interrupted:
            raise KeyboardInterrupt
        return rc

    def reactor_stop(self, reactor=None):
        if not reactor:
            reactor = self.get_reactor()
        self.flux_reactor_stop(reactor)
```

```python
import flux
import signal

h = flux.Flux()
h.timer_watcher_create (5., lambda h, w, x, y: h.reactor_stop()).start()
try:
    h.reactor()
except Exception as e:
    print ("exception ", e)
```

```console
$ flux python timer.py 
^C
Traceback (most recent call last):
  File "timer.py", line 7, in <module>
    h.reactor()
  File "/home/grondo/git/flux-core.git/src/bindings/python/flux/core/handle.py", line 193, in reactor
    raise KeyboardInterrupt
KeyboardInterrupt
```
